src/libraries/System.Net.Http/src/System/Net/Http/ReadOnlyMemoryContent.cs

```diff
@@ -18,6 +18,13 @@ public sealed class ReadOnlyMemoryContent : HttpContent
         protected override Task SerializeToStreamAsync(Stream stream, TransportContext? context) =>
             stream.WriteAsync(_content).AsTask();
 
+        protected override void SerializeToStream(Stream stream, TransportContext? context, CancellationToken cancellationToken)
+        {
+            // Last chance to check for timeout/cancellation, sync Stream API doesn't have any support for it.
+            cancellationToken.ThrowIfCancellationRequested();
+            stream.Write(_content.Span);
```

Comment:
The complete absence of cancellation/timeout support in these code paths is terrible. Some other components allow these operations to be cancelled by Disposing the stream.

Reply:
> The complete absence of cancellation/timeout support in these code paths is terrible. Some streams allow sync operations to be cancelled by Disposing the stream.

The main SendAsync method is registering with the CancellationToken to tear down the connection/stream should cancellation be requested at any time before all the response headers are received, and the implementation enforces that the response content is fully sent before that point.
https://github.com/dotnet/runtime/pull/34948/files#diff-75931c95adb5205b557555a5cd4f966eR336

Reply:
> and the implementation enforces that the response content is fully sent before that point.

You mean request content?

Glad to hear it. How about the default response buffering scenario?

Reply:
> You mean request content?

Sorry, yes, typo.  Request content.

> How about the default response buffering scenario?

I'd need to look; it's possible we rely on the cancellation token being passed around, e.g. into SerializeToStreamAsync.  If we don't already do something similar, we could for the sync path.

Reply:
It all falls apart right here: https://github.com/dotnet/runtime/pull/34948#discussion_r413328020

Reply:
In StreamContent?  That's not used in a response.

Reply:
Ah, that used to be the default response content type. Looks like it's HttpConnectionResponseContent now? That wasn't even updated in this PR. That means you end up in the base class around here? https://github.com/dotnet/runtime/pull/34948/files#diff-f00abc56209671d392efa6a8cf292483R326

Reply:
> That wasn't even updated in this PR.

If that's true, that sounds like a gap to be fixed.  I plan to review the PR tomorrow.

Reply:
Cool. Let me be clear about the affected call stack:
HttpClient.Send
HttpClient.FinishSend
HttpContent.LoadIntoBuffer
HttpContent.SerializeToStream
Stream.CopyTo - No cancellation

If you have a custom content type that can override SerializeToStream and register the token against the socket then that will help a lot.

Reply:
One thing we discussed re: cancellation token is to not plumb it everywhere but instead just dispose the socket. Considering we're only working with a few `Stream` implementations here, we should be able to get it working reasonably.

It falls apart if you use a custom transport (re: Bedrock) but I'd be fine throwing if someone uses a sync API with a custom transport, as those APIs won't support sync calls.

Reply:
Cancellation for buffered response content handled in `HttpContent.LoadIntoBuffer` via registering stream disposal in `CancellationToken` and then properly mapping exceptions. It's done the same way as in [`HttpConnection`](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs#L687).